core/block.go

```diff
@@ -14,23 +18,145 @@ type Block struct {
-	EventCount uint64
+	EventCount *felt.Felt
 	// A commitment to the events produced in this block
 	EventCommitment *felt.Felt
 	// The version of the StarkNet protocol used when creating this block
-	ProtocolVersion uint64
+	ProtocolVersion *felt.Felt
 	// Extraneous data that might be useful for running transactions
 	ExtraData *felt.Felt
 }
 
-func (b *Block) Hash() *felt.Felt {
-	// Todo: implement pedersen hash as defined here
-	// https://docs.starknet.io/documentation/develop/Blocks/header/#block_hash
-	return nil
+type BlockHashMetaInfo struct {
+	First07Block             uint64     // First block that uses the post-0.7.0 block hash algorithm
+	UnverifiableRange        []uint64   // Range of blocks that are not verifiable
+	FallBackSequencerAddress *felt.Felt // The sequencer address to use for blocks that do not have one
+}
+
+func getBlockHashMetaInfo(chain string) (*BlockHashMetaInfo, error) {
+	switch chain {
+	case "SN_MAIN":
+		fallBackSequencerAddress, _ := new(felt.Felt).SetString("0x021f4b90b0377c82bf330b7b5295820769e72d79d8acd0effa0ebde6e9988bc5")
+		return &BlockHashMetaInfo{
+			First07Block:             883,
+			UnverifiableRange:        []uint64{},
+			FallBackSequencerAddress: fallBackSequencerAddress,
+		}, nil
+	case "SN_GOERLI":
+		fallBackSequencerAddress, _ := new(felt.Felt).SetString("0x046a89ae102987331d369645031b49c27738ed096f2789c24449966da4c6de6b")
+		return &BlockHashMetaInfo{
+			First07Block:             47028,
+			UnverifiableRange:        []uint64{119802, 148428},
+			FallBackSequencerAddress: fallBackSequencerAddress,
+		}, nil
+	case "SN_GOERLI2":
+		fallBackSequencerAddress, _ := new(felt.Felt).SetString("0x046a89ae102987331d369645031b49c27738ed096f2789c24449966da4c6de6b")
+		return &BlockHashMetaInfo{
+			First07Block:             0,
+			UnverifiableRange:        []uint64{},
+			FallBackSequencerAddress: fallBackSequencerAddress,
+		}, nil
+	case "SN_INTEGRATION":
+		fallBackSequencerAddress, _ := new(felt.Felt).SetString("0x046a89ae102987331d369645031b49c27738ed096f2789c24449966da4c6de6b")
+		return &BlockHashMetaInfo{
+			First07Block:             110511,
+			UnverifiableRange:        []uint64{0, 110511},
+			FallBackSequencerAddress: fallBackSequencerAddress,
+		}, nil
+	default:
+		return nil, errors.New("unknown chain: " + chain)
+	}
+}
+
+// Block hash computation according to https://docs.starknet.io/documentation/develop/Blocks/header/#block_hash
+func (b *Block) Hash(chain string) (*felt.Felt, error) {
+	blockHashMetaInfo, err := getBlockHashMetaInfo(chain)
+	if err != nil {
+		return nil, err
+	}
+
+	unverifiableRange := blockHashMetaInfo.UnverifiableRange
+	if len(unverifiableRange) > 0 {
+		// Check if the block number is in the unverifiable range
+		if b.Number >= unverifiableRange[0] && b.Number <= unverifiableRange[len(unverifiableRange)-1] {
+			// If so, return unverifiable block error
+			return nil, errors.New("block " + strconv.FormatUint(b.Number, 10) + " is unverifiable")
```

Comment:
Slightly more idiomatic error message.
```suggestion
			return nil, errors.New("block is unverifiable: " + strconv.FormatUint(b.Number, 10))
```